fusestore/store/src/executor/action_handler.rs

```diff
@@ -103,158 +95,50 @@ impl ActionHandler {
         }
     }
 
-    async fn create_table(&self, act: CreateTableAction) -> Result<StoreDoActionResult, Status> {
-        let plan = act.plan;
-        let db_name = plan.db;
-        let table_name = plan.table;
-
-        info!("create table: {:}: {:?}", db_name, table_name);
-
-        let mut meta = self.meta.lock();
-
-        let options = common_arrow::arrow::ipc::writer::IpcWriteOptions::default();
-        let flight_data =
-            arrow_flight::utils::flight_data_from_arrow_schema(&plan.schema, &options);
-
-        let table = Table {
-            // the storage engine fills the id.
-            table_id: -1,
-            ver: -1,
-            schema: flight_data.data_header,
-            options: plan.options,
-
-            // TODO
-            placement_policy: vec![],
-        };
-
-        let cmd = CmdCreateTable {
-            db_name,
-            table_name,
-            table: Some(table),
-        };
-
-        let table_id = meta.create_table(cmd, plan.if_not_exists)?;
-
-        Ok(StoreDoActionResult::CreateTable(CreateTableActionResult {
-            table_id,
-        }))
-    }
-
-    async fn get_table(&self, act: GetTableAction) -> Result<StoreDoActionResult, Status> {
-        let db_name = act.db;
-        let table_name = act.table;
-
-        info!("create table: {:}: {:?}", db_name, table_name);
-
-        let mut meta = self.meta.lock();
-
-        let table = meta.get_table(db_name.clone(), table_name.clone())?;
-
-        let schema = Schema::try_from(&FlightData {
-            data_header: table.schema,
-            ..Default::default()
-        })
-        .map_err(|e| Status::internal(format!("invalid schema: {:}", e.to_string())))?;
-
-        let rst = StoreDoActionResult::GetTable(GetTableActionResult {
-            table_id: table.table_id,
-            db: db_name,
-            name: table_name,
-            schema: Arc::new(schema),
-        });
-
-        Ok(rst)
-    }
-
-    async fn drop_db(&self, act: DropDatabaseAction) -> Result<StoreDoActionResult, Status> {
-        let mut meta = self.meta.lock();
-        let _ = meta.drop_database(&act.plan.db, act.plan.if_exists)?;
-        Ok(StoreDoActionResult::DropDatabase(
-            DropDatabaseActionResult {},
-        ))
-    }
-
-    async fn drop_table(&self, act: DropTableAction) -> Result<StoreDoActionResult, Status> {
-        let mut meta = self.meta.lock();
-        let _ = meta.drop_table(&act.plan.db, &act.plan.table, act.plan.if_exists)?;
-        Ok(StoreDoActionResult::DropTable(DropTableActionResult {}))
+    async fn invoke<S, O, A, R>(&self, a: A, s: S) -> common_exception::Result<O>
+    where
+        A: Serialize + RequestFor<Reply = R>,
+        S: ReplySerializer<Output = O>,
+        Self: RequestHandler<A>,
+        R: Serialize,
+        <S as ReplySerializer>::Error: Into<ErrorCode>,
+    {
+        let r = self.handle(a).await?;
+        let v = s.serialize(r).map_err(Into::into)?;
+        Ok(v)
```

Comment:
Excellent abstraction!

Do you mind helping poor minds like me read the code easily?
It seems with this abstraction, I can never jump to a the actual handler anymore in Clion...

Reply:
- A:  the action,  for example, `UpsertKVAction`
           and the type of reply it requests should be encoded as `RequestFor`, e.g. `UpsertKVActionResult`
- R:  the reply which the action requests for, for example, `UpsertKVActionResult`
- S:  the serializer, which implements `ReplySerializer`, and  in response to encoding R into ReplySerializer::Out
          e.g. the JsonSer we are using:
~~~
struct JsonSer;
impl ReplySerializer for JsonSer {
    type Output = Vec<u8>;
    type Error = ErrorCode;
    fn serialize<T>(&self, v: T) -> Result<Self::Output, Self::Error>
    where T: Serialize {
        let v = serde_json::to_vec(&v).unwrap();
        Ok(v)
    }
}
~~~

- O:  equals ReplySerializer::Out


this method is for `ActionHandler::execute` to eliminates duplicated code, 
 > I can never jump to the actual handler anymore in Clion.
Yeah, that's a pity, if we do not mind the duplication, we can "restore" that, like 
~~~
    pub async fn execute<S, R>(&self, action: StoreDoAction, s: S) -> common_exception::Result<R>
    where
        S: ReplySerializer<Output = R>,
        <S as ReplySerializer>::Error: Into<ErrorCode>,
    {
        match action {
            // database
            // database
            StoreDoAction::CreateDatabase(a) => {
                let r = self.handle(a).await?;
                let v = s.serialize(r).map_err(Into::into)?;
                Ok(v)
            },
            StoreDoAction::GetDatabase(a) => {
                let r = self.handle(a).await?;
                let v = s.serialize(r).map_err(Into::into)?;
                Ok(v)
            }
            ...

            // table
        
            // part
            ...

            // general-purpose kv
           ....
        }
    }
~~~

Reply:
keep it IDE friendly (with some code duplications), commit 4038808